test/compilable/test14831.d

```diff
@@ -0,0 +1,60 @@
+// https://issues.dlang.org/show_bug.cgi?id=14831
+
+void main()
+{
+    {
+        int x;
```

Comment:
What would ever be the use of mangling a stack variable?

Reply:
Hmm, I missed that earlier in the review. You're right, local vars should not have a mangle.

Reply:
Passing local variables as template alias parameters, see the test case in #12236:

```d
void main()
{
    {
        int x = 1;
        print!x();
        pragma(msg, print!x.mangleof);
    }
    {
        int x = 2;
        print!x();
        pragma(msg, print!x.mangleof);
    }
}
```

Both instances of `print` receive the same mangling (`_D9test10619__T5printS_DQw4mainFZ1xiZQwMFNbNiZv`) without this patch.

Reply:
That's correct, though, isn't it?  The resultant code could use localized frames/closures if we know that the current scope isn't the top-level function scope.
```
void main()
{
    {
        __scoped_closure_main1 __scoped_closure;
        __scoped_closure.x = 1;
        print(&__scoped_closure);
    }
    {
        __scoped_closure_main2 __scoped_closure;
        __scoped_closure.x = 2;
        print(&__scoped_closure);
    }
}
```

Reply:
To suggest another left field idea, you could also detect local variable name collisions and put them in a union within the closure record.

So you end up with:
```
void main()
{
    struct __closure_main
    {
        union
        {
            int x;
            long x;
            real x;
            int x;
        }
        int y;
    }
    __closure_main __closure;
    {
        __closure.x = 1; // int x = 1;
        print(&__closure); // print!x();
    }
    {
        __closure.x = 2; // long x = 2;
        print(&__closure); // print!x();
    }
    {
        __closure.x = 3.4; // real x = 3.4;
        print(&__closure); // print!x();
    }
    {
        __closure.x = 5; // int x = 5;
        print(&__closure); // print!x();
    }
    __closure.y = 6; // int y = 6;
    print(&__closure); // print!y();
}
```

Reply:
Your suggestions assume that it's valid to merge both instances IIUC.
This might not be the case for such templates because one can inspect the alias parameter, e.g.

```d
void print(alias symbol)()
{
    import core.stdc.stdio : printf;
    printf("%d at line %d\n", symbol, __traits(getLocation, symbol)[1]);
}

void main()
{
    {
        int x = 1;
        print!x();
    }
    {
        int x = 2;
        print!x();
    }
}
```

Reply:
Hmm, but your suggestion of giving the local variable a different name would mean that you could not use `x` to look-up the value in a debugger, no?

Reply:
What about shadowing detection?  This would also bypass that as well?

Reply:
> Hmm, but your suggestion of giving the local variable a different name would mean that you could not use x to look-up the value in a debugger, no?

Debug information for local informations is not affected by this change (but is not correct atm)

Small test case:

```d
/*
https://issues.dlang.org/show_bug.cgi?id=10619

REQUIRED_ARGS: -g
PERMUTE_ARGS:
GDB_SCRIPT:
---
b gdb10619.d:26
b gdb10619.d:30
run

info locals
print x
continue

info locals
print x
continue
---
*/

void main()
{
    {
        int x = 1;
        print!x(); // first breakpoint
    }
    {
        int x = 2;
        print!x(); // second breakpoint
    }
}

void print(alias symbol)() {}
```

Stable:
```
Breakpoint 1 at 0x80b: file runnable/gdb10619.d, line 26.
Breakpoint 2 at 0x819: file runnable/gdb10619.d, line 30.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, D main () at runnable/gdb10619.d:26
26	        print!x();
x = 1
x = 0
$1 = 1

Breakpoint 2, D main () at runnable/gdb10619.d:30
30	        print!x();
x = 1
x = 2
$2 = 1
[Inferior 1 (process 466) exited normally]
```

This PR:
```
Breakpoint 1 at 0x80b: file runnable/gdb10619.d, line 26.
Breakpoint 2 at 0x819: file runnable/gdb10619.d, line 30.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, D main () at runnable/gdb10619.d:26
26	        print!x();
x = 1
x = 0
$1 = 1

Breakpoint 2, D main () at runnable/gdb10619.d:30
30	        print!x();
x = 1
x = 2
$2 = 1
[Inferior 1 (process 553) exited normally]
```

Reply:
> What about shadowing detection? This would also bypass that as well?

Shadowing detection is not affected by this PR, e.g.

```
void main()
{
    int x;
    {
        int x = 1; // Error: variable `x` is shadowing variable `gdb10619.main.x`
    }
}
````